tests/utils/test_dynamic_helper.py

```diff
@@ -0,0 +1,84 @@
+
+import importlib
+import sys
+from copy import copy
+from pathlib import Path
+
+import mock
+import pytest
+
+from volttron.utils.dynamic_helper import get_module, get_klass, get_subclasses
+
+# sys_path_with holds a sys.path that includes the path to
+# tests directory on the path.
+
+sys_path_with = copy(sys.path)
+sys_path_with.insert(0, str(Path(__file__).parent.parent))
+
+
+# Make sure these tests are first in file as after the path is specified and loaded
+# during the patch, the module under test will still be loaded.  I wasn't
+# able to figure out a clean way to unload modules.
+def test_get_module_import_error_raised():
+    with pytest.raises(ModuleNotFoundError):
+        get_module("module_for_testing.module_data")
+
+
+def test_get_klass_import_error_raised():
+    with pytest.raises(ModuleNotFoundError):
+        get_klass("module_for_testing.module_data", 'SecondaryClass')
+
+
+def test_get_module():
+    with mock.patch('sys.path', sys_path_with):
+        module_name = "module_for_testing.module_data"
+        actual_module = get_module("module_for_testing.module_data")
+        assert actual_module.__name__ == module_name
+
+
+def test_get_klass_str():
+    with mock.patch('sys.path', sys_path_with):
+        klass = get_klass("module_for_testing.module_data", 'SecondaryClass')
+        assert klass.__name__ == 'SecondaryClass'
+
+
+def test_get_klass_mod():
+    with mock.patch('sys.path', sys_path_with):
+        module = importlib.import_module("module_for_testing.module_data")
+        klass = get_klass(module, "MyHelperClass")
+        assert klass.__name__ == 'MyHelperClass'
+
+
+def test_get_klass_attribute_error_raised():
+    with pytest.raises(AttributeError) as err:
+        get_klass("module_for_testing.module_data", 'RandomClass')
+
+
+def test_get_subclasses_value_error_raised():
+    with mock.patch('sys.path', sys_path_with):
+        with pytest.raises(ValueError) as err:
+
+            module = importlib.import_module("module_for_testing.module_no_classes_data")
+            klass = getattr(importlib.import_module("module_for_testing.module_data"),
+                            "MyHelperClass")
+            get_subclasses(module, klass)
+
+
+def test_get_subclasses_returns_exactly_one_class():
+    with mock.patch('sys.path', sys_path_with):
+        module = importlib.import_module("module_for_testing.module_data")
+        klass = getattr(module, "MyHelperClass")
+        subclasses = get_subclasses(module, klass)
+        assert len(subclasses) == 1
+        assert subclasses[0].__name__ == 'SecondaryClass'
+
+
+def test_get_subclasses_returns_all_subclasses():
+    with mock.patch('sys.path', sys_path_with):
+        module = importlib.import_module("module_for_testing.module_data")
+        klass = getattr(module, "MyHelperClass")
+        subclasses = get_subclasses(module, klass, return_all=True)
+        assert len(subclasses) == 2
+        assert subclasses[0].__name__ == 'SecondaryClass'
```

Comment:
I wonder if the order of these is always the same?  I think so.

Reply:
I think so too. But I'll check.